File: articles/biomolecules-chemical-basis-of-life/fortran/biomolecule_kernel.f90
! Compact biomolecular numerical kernel in Fortran.

program biomolecule_kernel
  implicit none

  real :: substrate
  real :: vmax
  real :: km
  real :: velocity
  real :: ligand
  real :: kd
  real :: fraction_bound
  real :: diffusion_coefficient
  real :: concentration_gradient
  real :: flux
  integer :: monomer_count
  real :: mean_monomer_mass
  real :: water_loss_per_bond
  real :: polymer_mass

  substrate = 6.0
  vmax = 100.0
  km = 3.0
  velocity = (vmax * substrate) / (km + substrate)

  ligand = 8.0
  kd = 8.0
  fraction_bound = ligand / (kd + ligand)

  diffusion_coefficient = 2.0
  concentration_gradient = -0.8
  flux = -diffusion_coefficient * concentration_gradient

  monomer_count = 12
  mean_monomer_mass = 110.0
  water_loss_per_bond = 18.015
  polymer_mass = monomer_count * mean_monomer_mass - (monomer_count - 1) * water_loss_per_bond

  print *, "Michaelis-Menten velocity:", velocity
  print *, "Fraction bound:", fraction_bound
  print *, "Diffusive flux:", flux
  print *, "Estimated polymer mass:", polymer_mass
end program biomolecule_kernel

File: articles/biomolecules-chemical-basis-of-life/c/biomolecule_kernel.c
/*
 * Compact biomolecular kernel in C.
 */

#include <math.h>
#include <stdio.h>
#include <string.h>

double michaelis_menten(double substrate, double vmax, double km) {
    return (vmax * substrate) / (km + substrate);
}

double ligand_fraction_bound(double ligand, double kd) {
    return ligand / (kd + ligand);
}

double diffusive_flux(double diffusion_coefficient, double concentration_gradient) {
    return -diffusion_coefficient * concentration_gradient;
}

double polymer_mass_estimate(int monomer_count, double mean_monomer_mass, double water_loss_per_bond) {
    int bonds = monomer_count > 0 ? monomer_count - 1 : 0;
    return monomer_count * mean_monomer_mass - bonds * water_loss_per_bond;
}

double gc_content(const char *sequence) {
    int g = 0;
    int c = 0;
    int total = 0;

    for (size_t i = 0; i < strlen(sequence); i++) {
        if (sequence[i] == 'G') g++;
        if (sequence[i] == 'C') c++;
        if (sequence[i] == 'A' || sequence[i] == 'T' || sequence[i] == 'G' || sequence[i] == 'C') total++;
    }

    if (total == 0) {
        return NAN;
    }

    return (double)(g + c) / (double)total;
}

int main(void) {
    printf("velocity=%.6f\n", michaelis_menten(6.0, 100.0, 3.0));
    printf("fraction_bound=%.6f\n", ligand_fraction_bound(8.0, 8.0));
    printf("diffusive_flux=%.6f\n", diffusive_flux(2.0, -0.8));
    printf("polymer_mass=%.6f\n", polymer_mass_estimate(12, 110.0, 18.015));
    printf("gc_content=%.6f\n", gc_content("ATGCGCGTATTAACCGGTTAGCGCGATATCGCGTA"));

    return 0;
}
